capsules/src/net/tlv.rs

```diff
@@ -0,0 +1,1324 @@
+        let (offset, length) = dec_try!(buf, offset; decode_u8);
+        match tlv_type {
+            NetworkDataTlvType::Prefix => {
+                let (offset, domain_id) = dec_try!(buf, offset; decode_u8);
+                let (offset, prefix_length_bits) = dec_try!(buf, offset; decode_u8);
+                let mut prefix = [0u8; 3];
+                let offset = dec_consume!(buf, offset; decode_bytes_be, &mut prefix);
+                stream_done!(offset + length as usize,
+                             Some((NetworkDataTlv::Prefix {
+                                       domain_id: domain_id,
+                                       prefix_length_bits: prefix_length_bits,
+                                       prefix: prefix,
+                                       sub_tlvs: &buf[offset..offset + length as usize],
+                                   },
+                                   stable)))
+            }
+            NetworkDataTlvType::CommissioningData => {
+                let (offset, com_length) = dec_try!(buf, offset; decode_u8);
+                let mut com_data = [0u8; MAX_VALUE_LENGTH];
+                let offset = dec_consume!(buf, offset; decode_bytes_be, &mut com_data);
+                stream_done!(offset,
+                             Some((NetworkDataTlv::CommissioningData {
+                                       com_length: com_length,
+                                       com_data: com_data,
+                                   },
+                                   stable)))
+            }
+            NetworkDataTlvType::Service => {
+                let (offset, first_byte) = dec_try!(buf, offset; decode_u8);
+                let thread_enterprise_number = (first_byte >> 7) > 0;
+                let s_id = first_byte & 0b1111;
+                let (offset, s_enterprise_number) = dec_try!(buf, offset; decode_u32);
+                let (offset, s_service_data_length) = dec_try!(buf, offset; decode_u8);
+                let mut s_service_data = [0u8; MAX_VALUE_LENGTH];
+                let offset = dec_consume!(buf, offset; decode_bytes_be, &mut s_service_data);
+                stream_done!(offset + length as usize,
+                             Some((NetworkDataTlv::Service {
+                                       thread_enterprise_number: thread_enterprise_number,
+                                       s_id: s_id,
+                                       s_enterprise_number: s_enterprise_number,
+                                       s_service_data_length: s_service_data_length,
+                                       s_service_data: s_service_data,
+                                       sub_tlvs: &buf[offset..offset + length as usize],
+                                   },
+                                   stable)))
+            }
+            NetworkDataTlvType::NotPresent => stream_done!(offset, None),
+        }
+    }
+}
+
+#[repr(u8)]
+pub enum NetworkDataTlvType {
+    Prefix = 1,
```

Comment:
The annoying thing about these type numbers is that the ones missing here are filled by the Prefix and Service sub-TLVs. Added a comment explaining this.